cpp/src/gandiva/greedy_dual_size_cache.h

```diff
@@ -0,0 +1,139 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <list>
+#include <queue>
+#include <set>
+#include <unordered_map>
+#include <utility>
+
+#include "arrow/util/optional.h"
+#include "gandiva/base_cache.h"
+
+// modified cache to support evict policy using the GreedyDual-Size algorithm.
+namespace gandiva {
+// a particular LRU based cache which evicts the least recently used item
+// considering the different costs for each entry.
+template <class Key, class Value>
+class GreedyDualSizeCache : public BaseCache<Key, Value> {
+ // inner class to define the priority item
+ class PriorityItem {
+  public:
+   PriorityItem(uint64_t actual_priority,
+                uint64_t original_priority, Key key) : actual_priority(actual_priority),
+                original_priority(original_priority), cache_key(key) {}
+   // this ensure that the items with low priority stays in the beginning of the queue,
+   // so it can be the one removed by evict operation
+   bool operator<(const PriorityItem& other) const {
+     return this->actual_priority < other.actual_priority;
+   }
+   uint64_t actual_priority;
+   uint64_t original_priority;
+   Key cache_key;
+  };
+ public:
+  struct hasher {
+    template <typename I>
+    std::size_t operator()(const I& i) const {
+      return i.Hash();
+    }
+  };
+  // a map from 'key' to a pair of Value and a pointer to the priority value
+  using map_type = std::unordered_map<
+      Key, std::pair<Value, typename std::set<PriorityItem>::iterator>, hasher>;
+
+  explicit GreedyDualSizeCache(size_t capacity) : BaseCache<Key, Value>(capacity) {
+    this->inflation_cost_ = 0;
+  }
+
+  GreedyDualSizeCache<Key, Value>() : BaseCache<Key, Value>() {}
+
+  ~GreedyDualSizeCache() = default;
+
+  size_t size() const override { return map_.size(); }
+
+  size_t capacity() const override { return this->cache_capacity_; }
+
+  bool empty() const override { return map_.empty(); }
+
+  bool contains(const Key& key) override { return map_.find(key) != map_.end(); }
+
+  void insert(const Key& key, const Value& value) override {
+    insert(key, value, 0);
+  }
+
+  void insert(const Key& key, const Value& value, const uint64_t priority) {
```

Comment:
it should just take key and value with value as struct containing the cost

Reply:
I sent my opinion about the usage of the "Struct Object" on the comment before. Please, take a look if you can

Reply:
That was a different class (cache). There you are creating the struct and calling the other insert method on this class